always-on-source/AlwaysOnDisplay.mc

```diff
@@ -1,201 +1,212 @@
-		dc.drawText(
-			dc.getWidth() - mDataLeft,
-			y,
-			mBatteryFont,
-			battery.format(INTEGER_FORMAT) + "%",
-			Graphics.TEXT_JUSTIFY_RIGHT | Graphics.TEXT_JUSTIFY_VCENTER
-		);
-	}
+using Toybox.WatchUi as Ui;
+using Toybox.System as Sys;
+using Toybox.Application as App;
+using Toybox.Time;
+using Toybox.Time.Gregorian;
+
+// Draw time, line, date, battery.
+// Combine stripped down versions of ThickThinTime and DateLine.
+// Change vertical offset every minute to comply with burn-in protection requirements.
+class AlwaysOnDisplay extends Ui.Drawable {
+
+	private var mBurnInYOffsets;
+	private var mHoursFont, mMinutesFont, mSecondsFont, mDateFont, mBatteryFont;
+
+	// Wide rectangle: time should be moved up slightly to centre within available space.
+	private var mAdjustY = 0;
+
+	private var mTimeY;
+	private var mLineY;
+	private var mLineWidth;
+	private var mLineStroke;
+	private var mDataY;
+	private var mDataLeft;
+
+	private var AM_PM_X_OFFSET = 2;
+
+	private var mDayOfWeek;
+	private var mDayOfWeekString;
+
+	private var mMonth;
+	private var mMonthString;
+
+	function initialize(params) {
+		Drawable.initialize(params);
+
+		mBurnInYOffsets = params[:burnInYOffsets];
+
+		if (params[:adjustY] != null) {
+			mAdjustY = params[:adjustY];
+		}
+
+		if (params[:amPmOffset] != null) {
+			AM_PM_X_OFFSET = params[:amPmOffset];
+		}
+
+		mTimeY = params[:timeY];
+		mLineY = params[:lineY];
+		mLineWidth = params[:lineWidth];
+		//mLineStroke = params[:lineStroke];
+		mDataY = params[:dataY];
+		mDataLeft = params[:dataLeft];
+
+		mHoursFont = Ui.loadResource(Rez.Fonts.AlwaysOnHoursFont);
+		mMinutesFont = Ui.loadResource(Rez.Fonts.AlwaysOnMinutesFont);
+		mSecondsFont = Ui.loadResource(Rez.Fonts.AlwaysOnSecondsFont);
+		mBatteryFont = Ui.loadResource(Rez.Fonts.AlwaysOnBatteryFont);
+
+		var rezFonts = Rez.Fonts;
+		var resourceMap = {
+			"ZHS" => rezFonts.AlwaysOnDateFontOverrideZHS,
+			"ZHT" => rezFonts.AlwaysOnDateFontOverrideZHT,
+			"RUS" => rezFonts.AlwaysOnDateFontOverrideRUS
+		};
+
+		// Unfortunate: because fonts can't be overridden based on locale, we have to read in current locale as manually-specified
+		// string, then override font in code.
+		var dateFontOverride = Ui.loadResource(Rez.Strings.DATE_FONT_OVERRIDE);
+		var dateFont = (resourceMap.hasKey(dateFontOverride)) ? resourceMap[dateFontOverride] : rezFonts.AlwaysOnDateFont;
+		mDateFont = Ui.loadResource(dateFont);
+	}
+	
+	function draw(dc) {
+
+		// TIME.
+		var clockTime = Sys.getClockTime();
+		var formattedTime = App.getApp().getFormattedTime(clockTime.hour, clockTime.min);
+		formattedTime[:amPm] = formattedTime[:amPm].toUpper();
+
+		// Change vertical offset every minute.
+		var burnInYOffset = mBurnInYOffsets[clockTime.min % mBurnInYOffsets.size()] + (clockTime.min - 30);
+
+		var hours = formattedTime[:hour];
+		var minutes = formattedTime[:min];
+		var amPmText = formattedTime[:amPm];
+		var colon = ":"; // SG Addition

```

Comment:
I'm afraid I shouldn't allow the colon and font change to the always-on display: the bold hours/light minutes display is a defining characteristic of Crystal's display, so I'd prefer to keep this as is. The time display should be consistent between regular, and always-on.

For reference, note that this change also affects the newly-supported `venusq` and `venusqm` watches, which are currently broken:

![image](https://user-images.githubusercontent.com/6435959/103681895-faacc680-4f7f-11eb-8928-ea5d05711207.png)

The line endings have changed throughout this file.

Would you mind reversing changes to this file, and also to `resources-round-390x390/fonts/fonts.xml`?

We could consider making this change settings-based in future, although I'm not sure how to resolve the inconsistency with the regular time display, as there's no way a colon would fit for some devices.

Reply:
Sorry, I've just noticed that the reason for making the always-on changes was to reduce battery drain. Is the saving due to using fewer fonts?

Reply:
I've checked all the 'AlwaysOnHoursFont' and made sure the ':' is present.